### PR TITLE
GCR Krylov Solver

### PR DESCRIPTION
**Pull request by kho19**

### CODE REVIEW

**Comment by upsj** [comment]
label!

**Comment by upsj** [comment]
check-format!

**Comment by MarcelKoch** [comment]
Thanks @kho19 for your contribution. Before I get to reviewing, do you mind posting a literature reference for the algorithm you've implemented? It would also be nice to have that reference in the code documentation.

**Review by yhmtsai — COMMENTED** [review]
Could you add the description for this pull requirement?

**Comment by kho19** [comment]
GCR was based on the description in _Iterative methods for sparse linear systems_ by Yousef Saad. I'll include this in the documentation.

File: test/solver/gcr_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        residual = gen_mtx(m, nrhs);
+        A_residual = gen_mtx(m, nrhs);
+        p_bases = gen_mtx(m * (gko::solver::gcr_default_krylov_dim + 1), nrhs);
+        p = gen_mtx(m, nrhs);
+        Ap_bases = gen_mtx(m * (gko::solver::gcr_default_krylov_dim + 1),
+                           nrhs);
+        Ap = gen_mtx(m, nrhs);
+        alpha = gen_mtx(1, nrhs);
+        Ap_norm = gen_mtx(1, nrhs);
+
```
Comment (MarcelKoch) [inline test/solver/gcr_kernels.cpp:129]:
```suggestion
```

File: test/solver/gcr_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    std::unique_ptr<Mtx> d_A_residual;
+    std::unique_ptr<Mtx> d_p_bases;
+    std::unique_ptr<Mtx> d_p;
+    std::unique_ptr<Mtx> d_Ap_bases;
+    std::unique_ptr<Mtx> d_Ap;
+    std::unique_ptr<Mtx> d_alpha;
+    std::unique_ptr<Mtx> d_Ap_norm;
+    gko::Array<gko::stopping_status> d_stop_status;
+    gko::Array<gko::size_type> d_final_iter_nums;
+
```
Comment (MarcelKoch) [inline test/solver/gcr_kernels.cpp:189]:
```suggestion
```

**Review by MarcelKoch — CHANGES_REQUESTED** [review]
The implementation looks good. I've remarked mostly some formatting stuff.

The largest issue I have is that there are no reference tests. This needs to be fixed since you compare the other implementations to the reference one.

Also, I think this needs to be rebased. I'm not sure if the changes in the gmres code are due to this PR or the gmres refactoring one, so I skipped all of them.

**Pull request closed without merging**